controllers/url_controller.js

```diff
@@ -0,0 +1,62 @@
+import Url from "../models/Url.js";
+
+const shorten_url = async (req, res, next) => {
+    try{
+        let { original_url } = req.body;
+
+        if (!original_url) {
+            return res.status(400).send({ error: 'A URL is required.' });
+        }
+
+        // Remove all whitespace from the URL and trim it
+        original_url = original_url.replace(/\s+/g, '').trim();
+
+        // Validate the URL
+        try {
+            new URL(original_url);
+        } catch (e) {
+            return res.status(400).send({ error: 'Invalid URL format.' });
+        }
+
+        const update = { original_url };
+        const options = { new: true, upsert: true, setDefaultsOnInsert: true };
+    
+        // Use findOneAndUpdate with upsert option to reduce database calls
+        const url = await Url.findOneAndUpdate({ original_url }, update, options);
+
+        return res.send({ 
+            original_url: url.original_url, 
+            short_url: `${process.env.DOMAIN}${url.short_url }`
+        }); 
+    }
+    catch(error){
+        return res.status(500).send({ error: 'Server error' });
+    }
+};
+
+const url_redirect = async (req, res, next) => {
+    try {
+
+        const { url } = req.params;
+        
+        if(!url){
+            return res.status(400).send({ error: 'A short url is required.' });
```

Comment:
This if(!url) check will never get triggered in my current code because the route for this handler is /':url' which means that if the url parameter is missing it will redirect to the default '/' route and never trigger this 'url_redirect' function. Regardless this might be seemed as defensive programming and in the case that my route prefixes change, this will be needed.